frontend: Add create namespace ui

vyncent-t opened the pull request:
# Add Create Namespace Feature to UI

fixes #1705 
fixes #2055 

## Description
This PR introduces a new feature to the UI that allows users to create namespaces directly from the namespaces list view. A 'Create' button (represented by a + sign) has been added next to the title of the namespaces list view. Upon clicking this button, users are presented with a dialog box where they can input the name for the new namespace. Submitting this by clicking the 'Create' button will create the new namespace.

## Changes
- [x] Added a 'Create' button (+ sign) next to the title of the namespaces list view.
- [x] Implemented a dialog box that prompts the user to enter the name of the new namespace.
- [x] Added functionality to create a new namespace upon clicking the 'Create' button in the dialog box.

## Verification
- [x] Tested the 'Create' button to ensure it triggers the dialog box as expected.
- [x] Verified the dialog box accepts input and allows for the creation of a new namespace.
- [x] Confirmed that the new namespace appears in the namespaces list view upon creation.

## Screenshots
![image](https://github.com/headlamp-k8s/headlamp/assets/78232183/432c3687-b4f5-416e-bf46-5e793efb3877)
![image](https://github.com/headlamp-k8s/headlamp/assets/78232183/e3b13760-62b3-428b-822c-b4372b2b6a2e)

vyncent-t commented:
- push to fix build failures and translations

knrt10 commented on frontend/src/i18n/locales/de/translation.json:
```diff
@@ -303,6 +303,7 @@
   "Default Request": "Standardanforderung",
   "Max": "Max",
   "Min": "Min",
+  "Create Namespace": "",
```
```suggestion
  "Create Namespace": "Namensraum erstellen",
```

knrt10 commented on frontend/src/i18n/locales/es/translation.json:
```diff
@@ -303,6 +303,7 @@
   "Default Request": "Solicitud por defecto",
   "Max": "Máx.",
   "Min": "Mín.",
+  "Create Namespace": "",
```
```suggestion
  "Create Namespace": "Crear espacio de nombres",
```

knrt10 commented on frontend/src/i18n/locales/pt/translation.json:
```diff
@@ -303,6 +303,7 @@
   "Default Request": "Pedido por defeito",
   "Max": "Máx.",
   "Min": "Mín.",
+  "Create Namespace": "",
```
```suggestion
  "Create Namespace": "Criar Namespace",
```

knrt10 reviewed (commented):
Some suggestions for translations

joaquimrocha reviewed (commented):
Thanks Vincent.
Can you move the dialog/button into its own module? So it's more isolated and also in case we want to show it in a different view.

For consistency with the rest of the UX, I think the create button should trigger a cancellable action (like the Editor's apply / Delete / Scale / etc.).

joaquimrocha reviewed (changes requested):
Req changes.

vyncent-t commented:
- last push removes notes

joaquimrocha commented on frontend/src/components/namespace/CreateNamespaceDialog.tsx:
Call it CreateNamespaceButton.tsx

joaquimrocha commented on frontend/src/components/namespace/CreateNamespaceDialog.tsx:
```diff
@@ -0,0 +1,92 @@
+      description={t('translation|Create')}
+      icon={'mdi:plus-circle'}
+      onClick={() => {
+        toggleFunc();
```
The logic for toggling the dialog for this button can be done within this button. See the DeleteButton.
We also need it visible only if the user can create namespaces (see use of AuthVisible).

vyncent-t commented:
- last push is still needing some fixes, when creating a namespace item it is almost instant at this current implementation.

maybe artificially adding a timer would allow it to be slow enough to react to a quit action?

joaquimrocha commented on frontend/src/components/namespace/CreateNamespaceButton.tsx:
```diff
@@ -0,0 +1,126 @@
+            successMessage: t('Created namesapce {{ itemName }}.', { itemName }),
+            errorMessage: t('Error creating namespace {{ itemName }}.', { itemName }),
+            cancelUrl: location.pathname,
+            // startUrl: item!.getListLink(),
+            // errorUrl: item!.getListLink(),
+            // ...options,
```
Please remove this.

vyncent-t replied:
sorry this was wip and didn't mean to have it reviewed yet!

joaquimrocha commented on frontend/src/components/namespace/CreateNamespaceButton.tsx:
```diff
@@ -0,0 +1,126 @@
+    async function createNamesapce(newNamespaceData: any) {
+      try {
+        const response = await post('/api/v1/namespaces', newNamespaceData);
+        console.log('response', response);
```
Please remove this.

joaquimrocha commented on frontend/src/components/namespace/CreateNamespaceButton.tsx:
```diff
@@ -0,0 +1,126 @@
+
+    async function createNamesapce(newNamespaceData: any) {
+      try {
+        const response = await post('/api/v1/namespaces', newNamespaceData);
```
We need to pass the cluster here. And this function can only be called after the grace period. i.e. it should be passed to the `clusterAction`.

vyncent-t replied:
post logic currently handles setting the cluster

joaquimrocha replied:
Here is the problem: the post handles setting the cluster, but post is called only after the grace period, and by then the user may have changed to another cluster (and that's the one the post picks up). This is why we had to add the cluster as an arg to these functions. In the future we should have the cluster as a mandatory argument to avoid this.

joaquimrocha commented on frontend/src/components/namespace/CreateNamespaceButton.tsx:
```diff
@@ -0,0 +1,126 @@
+    // need this for the item name?
+    const itemName = newNamespaceData.metadata.name;
+
+    // need a callback so just made one to use
+    const callback = (itemName: string) => {
+      console.log(`Callback called with ${itemName}`);
+    };
```
The callback should be the namespace creation (the post).

vyncent-t replied:
Created a more unique name

joaquimrocha commented on frontend/src/components/namespace/CreateNamespaceButton.tsx:
```diff
@@ -0,0 +1,126 @@
+    setNamespaceDialogOpen(namespaceDialogOpen => !namespaceDialogOpen);
+  }
+
+  console.log('kubeNamespace', kubeNamespace);
```
Please remove this.

joaquimrocha commented on frontend/src/components/namespace/CreateNamespaceButton.tsx:
```diff
@@ -0,0 +1,126 @@
+        const response = await post('/api/v1/namespaces', newNamespaceData);
+        console.log('response', response);
+
+        toggleCreateNamespaceVisibility();
```
I think we can avoid future discrepancies by just explicitly hiding the dialog, instead of toggling the visibility (which I understands hides it, but maybe not in the future).

vyncent-t replied:
Fixed it!

joaquimrocha commented on frontend/src/components/namespace/CreateNamespaceButton.tsx:
```diff
@@ -0,0 +1,126 @@
+      } catch (error) {
+        console.error('Error creating namespace', error);
+
+        toggleCreateNamespaceVisibility();
```
Explicitly show the dialog please.

vyncent-t replied:
sorry this was wip and didn't mean to have it reviewed yet!

joaquimrocha commented on frontend/src/components/namespace/CreateNamespaceButton.tsx:
```diff
@@ -0,0 +1,126 @@
+      }
+    }
+
+    createNamesapce(newNamespaceData);
```
?

vyncent-t replied:
sorry this was wip and didn't mean to have it reviewed yet!

joaquimrocha commented on frontend/src/components/namespace/CreateNamespaceButton.tsx:
```diff
@@ -0,0 +1,126 @@
+  const [namespaceName, setNamespaceName] = useState('');
+  const [namespaceDialogOpen, setNamespaceDialogOpen] = useState(false);
+  const dispatchCreateEvent = useEventCallback(HeadlampEventType.CREATE_RESOURCE);
+
```
Remove this newline.

joaquimrocha commented on frontend/src/components/namespace/CreateNamespaceButton.tsx:
```diff
@@ -0,0 +1,126 @@
+  }
+
+  return (
+    <AuthVisible item={kubeNamespace} authVerb="create" namespace="default">
```
Namespaces aren't namespaced. So the namespace argument is not needed here.
I am not sure the Namespace class can be used here like this either. Maybe AuthVisible needs changes to accommodate what we need.

joaquimrocha commented on frontend/src/components/namespace/CreateNamespaceButton.tsx:
```diff
@@ -0,0 +1,126 @@
+        <DialogTitle>{t('translation|Create Namespace')}</DialogTitle>
+        <DialogContent>
+          <Box component="form">
+            <Typography>Please enter the name of the new namespace</Typography>
```
Needs translation.

vyncent-t replied:
would this be correct?
`<Typography>{t('translation|Please enter the name of the new namespace')}</Typography>`

joaquimrocha commented on frontend/src/components/namespace/CreateNamespaceButton.tsx:
```diff
@@ -0,0 +1,126 @@
+          <Button
+            onClick={() => {
+              toggleCreateNamespaceVisibility();
+              dispatchCreateEvent({
+                status: EventStatus.CONFIRMED,
+              });
```
This is reporting that the namespace creation was confirmed when in fact this button cancels it.

joaquimrocha commented on frontend/src/i18n/locales/de/translation.json:
```diff
@@ -303,6 +303,7 @@
   "Default Request": "Standardanforderung",
   "Max": "Max",
   "Min": "Min",
+  "Create Namespace": "Namensraum erstellen",
```
Namespace in glossary is translated as "Namespace" since it's a technical term. So in here the translation is not correct.
Please leave it empty or update it to using Namespace.

vyncent-t replied:
change to `"Create Namespace": ""` ?

joaquimrocha commented on frontend/src/components/namespace/CreateNamespaceButton.tsx:
```diff
@@ -0,0 +1,106 @@
+    setNamespaceDialogOpen(false);
+    dispatch(
+      clusterAction(() => namespaceRequest(), {
+        startMessage: t('Creating namespace {{ itemName }}…', { namespaceName }),
+        cancelledMessage: t('Cancelled creation of {{ itemName }}.', { namespaceName }),
+        successMessage: t('Created Namespace {{ itemName }}.', { namespaceName }),
+        errorMessage: t('Error creating namespace {{ itemName }}.', { namespaceName }),
```
The `itemName` doesn't match any key in the passed object. And testing shows:
![Screenshot showing the cancel message without the name of the namespace](https://github.com/headlamp-k8s/headlamp/assets/1029635/3a54edb0-6bc7-4a7f-9f52-c3032fb8cf9a)

vyncent-t replied:
fixed! let me know if this usage is okay, since the variable being used a custom name by the user not sure if it would need a direct translation back to them

joaquimrocha commented on frontend/src/components/namespace/CreateNamespaceButton.tsx:
```diff
@@ -0,0 +1,106 @@
+      clusterAction(() => namespaceRequest(), {
+        startMessage: t('Creating namespace {{ itemName }}…', { namespaceName }),
+        cancelledMessage: t('Cancelled creation of {{ itemName }}.', { namespaceName }),
+        successMessage: t('Created Namespace {{ itemName }}.', { namespaceName }),
```
Do not capitalize the N here. So we are consistent with the rest of the messages.

joaquimrocha commented on frontend/src/components/namespace/CreateNamespaceButton.tsx:
```diff
@@ -0,0 +1,106 @@
+        cancelledMessage: t('Cancelled creation of {{ itemName }}.', { namespaceName }),
+        successMessage: t('Created Namespace {{ itemName }}.', { namespaceName }),
+        errorMessage: t('Error creating namespace {{ itemName }}.', { namespaceName }),
+        cancelUrl: location.pathname,
```
Since the URL doesn't change. You can use the `cancelCallback` argument instead for bringing up the namespace creation dialog on cancel.

vyncent-t replied:
Fixed! using the cancelCallback to set dialog to open

joaquimrocha commented on frontend/src/components/namespace/List.tsx:
```diff
@@ -86,12 +87,15 @@ export default function NamespacesList() {
-      }}
-      {...resourceTableProps}
-    />
+    <>
```
This is not needed AFAIU.

vyncent-t replied:
The headerProps for titleSideActions is what lets the button render neatly next to the page title

sniok replied:
I think he meant the Fragment `<></>` that wraps ResourceListView

vyncent-t replied:
removed the `<></>`

joaquimrocha reviewed (changes requested):
Left a few more comments.

vyncent-t commented:
rebased and changed Namespace to namespace

vyncent-t commented:
Fixed some changes and should be ready for review again

sniok commented on frontend/src/components/namespace/CreateNamespaceButton.tsx:
```diff
@@ -0,0 +1,118 @@
+        <DialogTitle>{t('translation|Create Namespace')}</DialogTitle>
+        <DialogContent>
+          <Box component="form" style={{ width: '20vw', maxWidth: '20vw' }}>
+            <TextField
```
It'd be nice to have some validation for the namespace name, like max characters and allowed characters

vyncent-t replied:
I tried using some special chars and it doesnt seem to create the namespace with special chars, the length could be long though I tried that also. What would be a good word length for names?

sniok replied:
According to the docs it should be a valid DNS name, more on that here
https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-label-names https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-label-names

vyncent-t replied:
Fixed!

joaquimrocha commented on frontend/src/components/namespace/CreateNamespaceButton.tsx:
```diff
@@ -0,0 +1,163 @@
+
+  function createNewNamespace() {
+    function validateNamespaceName(namespaceName: string) {
+      const namespaceRegex = /^[a-z]([-a-z0-9]*[a-z0-9])?$/;
```
We have a function for this at SettingsCluster.tsx. We should add this as a static method to the Namespace class so we can reuse it when needed.

vyncent-t replied:
fixed!

joaquimrocha commented on frontend/src/components/namespace/CreateNamespaceButton.tsx:
```diff
@@ -0,0 +1,163 @@
+    function validateNamespaceName(namespaceName: string) {
+      const namespaceRegex = /^[a-z]([-a-z0-9]*[a-z0-9])?$/;
+
+      // Check if the namespace name is too long
+      if (namespaceName.length > 63) {
+        alert(t('translation|Invalid namespace name: name must be 63 characters or shorter'));
+        return false;
+      }
+
+      // Check if the namespace name starts with a alphanumeric character
+      const startsAlphanumeric = /^[a-z0-9]/.test(namespaceName);
+      if (!startsAlphanumeric) {
+        alert(
+          t('translation|Invalid namespace name: name must start with an alphanumeric character')
+        );
+        return false;
+      }
+
+      // Check if the namespace name ends with an alphanumeric character
+      const endsAlphanumeric = /[a-z0-9]$/.test(namespaceName);
+      if (!endsAlphanumeric) {
+        alert(
+          t('translation|Invalid namespace name: name must end with an alphanumeric character')
+        );
+        return false;
+      }
+
+      // Check if the namespace name contains only lowercase letters, numbers, or hyphens
+      const validChars = namespaceRegex.test(namespaceName);
+      if (!validChars) {
+        alert(
+          t(
+            'translation|Invalid namespace name: name must consist of lower case alphanumeric characters or -'
+          )
+        );
+        return false;
+      }
+
+      return true;
+    }
```
See my comment about the function we have in SettingsCluster.tsx.

vyncent-t replied:
fixed!

joaquimrocha commented on frontend/src/components/namespace/CreateNamespaceButton.tsx:
```diff
@@ -0,0 +1,163 @@
+    async function namespaceRequest() {
+      try {
+        const response = await post('/api/v1/namespaces', newNamespaceData, true, {
+          cluster: `${cluster}`,
```
You shouldn't use a string literal here (because if cluster is `null`, it will result in a cluster called `"null"`). You can use `getCluster() || ''` when you declare the var up there.

vyncent-t replied:
fixed!

joaquimrocha commented on frontend/src/components/namespace/CreateNamespaceButton.tsx:
```diff
@@ -0,0 +1,163 @@
+        const response = await post('/api/v1/namespaces', newNamespaceData, true, {
+          cluster: `${cluster}`,
+        });
+        return response;
```
You can return without having to declare a var in between (`return await post...`)

joaquimrocha commented on frontend/src/i18n/locales/de/translation.json:
```diff
@@ -315,6 +315,15 @@
   "Max": "Max",
   "Min": "Min",
+  "Invalid namespace name: name must be 63 characters or shorter": "",
+  "Invalid namespace name: name must start with an alphanumeric character": "",
+  "Invalid namespace name: name must end with an alphanumeric character": "",
+  "Invalid namespace name: name must consist of lower case alphanumeric characters or -": "",
```
Can you reuse the string we already have for SettingsCluster? So we don't have to translate these? (We are missing the 63 chars one, so maybe keep that or adapt the new one)

vyncent-t replied:
fixed!

joaquimrocha commented on frontend/src/i18n/locales/fr/translation.json:
```diff
@@ -315,6 +315,15 @@
+  "Invalid namespace name: name must start with an alphanumeric character": "",
+  "Invalid namespace name: name must end with an alphanumeric character": "",
+  "Invalid namespace name: name must consist of lower case alphanumeric characters or -": "",
+  "Creating namespace {{  name  }}…": "",
+  "Cancelled creation of {{  name  }}.": "",
+  "Created namespace {{  name  }}.": "",
+  "Error creating namespace {{  name  }}.": "",
+  "Create Namespace": "",
```
I wonder if in the interest of not having to translate 5 strings per resource, we can reuse the string "Failed to create {{ kind }} {{ name }}." and similar.

joaquimrocha reviewed (changes requested):
Left a few notes. Please take a look.

vyncent-t commented:
Fixed start message to the correct use

illume commented on frontend/src/components/namespace/CreateNamespaceButton.tsx:
Can you please add a CreateNamespaceButton.stories.tsx for this?

illume commented on frontend/src/lib/k8s/namespace.ts:
```diff
@@ -13,6 +13,18 @@ class Namespace extends makeKubeObject<KubeNamespace>('namespace') {
+
+    // Validates that the namespace is a valid DNS-1123 label and returns a boolean.
+    // https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-label-names
+    const regex = new RegExp('^[a-z0-9]([-a-z0-9]*[a-z0-9])?$');
```
There is this rule which the function does not validate:
"contain at most 63 characters"

vyncent-t replied:
fixed!

illume commented on frontend/src/lib/k8s/namespace.ts:
```diff
@@ -13,6 +13,18 @@ class Namespace extends makeKubeObject<KubeNamespace>('namespace') {
+
+  static isValidNamespaceFormat(namespace: string) {
+    // We allow empty strings just because that's the default value in our case.
+    if (!namespace) {
```
Can you please add a unit test for this function?

vyncent-t replied:
added a namespace.test.ts for unit testing but I am hitting a blocker with this error

![image](https://github.com/headlamp-k8s/headlamp/assets/78232183/9fe783ee-c6c3-4b7d-8438-2b76d07f7b77)


The error says the makeKubeObject is not a function, yet it is one and the import for it should be fine on the normal namespace.ts, what is even more weird is that I cannot seem to find how or why its using the Role extends as I do not see anything that Namespace class extends directly off of any Role usage? other than searching project wide for Role or ClusterRole extensions?

illume commented on frontend/src/lib/k8s/namespace.ts:
```diff
@@ -13,6 +13,18 @@ class Namespace extends makeKubeObject<KubeNamespace>('namespace') {
     return this.jsonData!.status;
   }
+
+  static isValidNamespaceFormat(namespace: string) {
```
Please add docs for this function?

vyncent-t replied:
Fixed!

illume reviewed (changes requested):
Nice work.

I left a few notes.

Can you please also rebase this against main? It's been a while.

illume commented:
The build Container and test job failed so I restarted it in case it was flaky.

vyncent-t commented:
### Prev push

- fixed (for now) the error when using namespace.test.ts would cause `npm run test` to fail due to "TypeError: makeKubeObject is not a function"

joaquimrocha commented on frontend/src/lib/k8s/index.test.ts:
```diff
@@ -264,3 +265,24 @@ describe('Test class namespaces', () => {
+ * note: this test will not work correctly if used in its own environment (i.e. namespace.test.ts)
+ */
+describe('Namespace testing', () => {
+  describe('test working isValidNamespaceFormat', () => {
```
You should test more cases:
empty name;
names with capitals;

joaquimrocha commented on frontend/src/components/namespace/CreateNamespaceButton.tsx:
```diff
@@ -0,0 +1,129 @@
+
+  function createNewNamespace() {
+    if (!Namespace.isValidNamespaceFormat(namespaceName)) {
+      alert(
```
For telling the user about the error, use the error + helper text functionality of the text field as we do in the cluster settings.

joaquimrocha reviewed (changes requested):
Thanks Vincent. I left the comments we had also talked about during our call.

vyncent-t commented:
### Push
- fixes review comments

- [ ] current push does not allow users to input empty string for a namespace name, need to implement a way to use a default name

vyncent-t commented:
- [ ] create an error for if namespace already exist

vyncent-t commented:
Currently trying to work around an issue for the storybook component, it seems that the storybook compile is running into the same circular dependency error that the namespace.test.ts was hitting (in short it would fail if we tried to create a unit test for the namespace functions in its own .test env) 

![image](https://github.com/headlamp-k8s/headlamp/assets/78232183/356b4739-65e4-4a76-a11e-5336c9046af2)

vyncent-t commented:
- [ ] need to fix being able to use enter key to create

vyncent-t commented:
### Push note
- [x] Fixed page reloading when pressing enter when writing name
- [ ] story component still broken

vyncent-t commented:
### push note

- Removed the <AuthVisible> comp around the create button as it is not working correctly with the params (example: this doesn't prevent the button from rendering)
![image](https://github.com/headlamp-k8s/headlamp/assets/78232183/3f03d04f-5246-411f-a5cc-e66898c6346b)

- [x] Fixed story comp not rendering

vyncent-t commented:
### prev push
- [x] Adds working story components

vyncent-t commented:
The build storybook error is likely a bug due to the storybook vite build on the publisher side https://github.com/storybookjs/storybook/issues/26685

I may just remove the test for now until this is fixed for vite

vyncent-t commented:
I was able to find a fix, seems the latest update isn't vite compatible so rolled back to the previous version

- storybook/addons-interactions `7.6.20`

joaquimrocha commented on frontend/src/components/namespace/CreateNamespaceButton.tsx:
```diff
@@ -0,0 +1,167 @@
+        if (errorMessage.includes('already exists')) {
+          setNamespaceDialogOpen(true);
+          setNamespaceNameValid(false);
+          setNameHelperMessage(t('translation|Namespace name already exists.'));
```
This text sounds a bit odd. Maybe just "Namespace already exists" or "A namespace with this name already exists"

joaquimrocha commented on frontend/src/components/namespace/CreateNamespaceButton.tsx:
```diff
@@ -0,0 +1,167 @@
+      } catch (error) {
+        console.error('Error creating namespace', error);
+        const errorMessage = error instanceof Error ? error.message : String(error);
+        if (errorMessage.includes('already exists')) {
```
Checking for substrings in error messages is not a good practice (as those are not standard and may change).
Can you check for the error code instead?

joaquimrocha commented on frontend/src/components/namespace/CreateNamespaceButton.tsx:
```diff
@@ -0,0 +1,167 @@
+
+    async function namespaceRequest() {
+      try {
+        return await post('/api/v1/namespaces', newNamespaceData, true, {
```
Please use `Namespace.apiEndpoint.post`.

joaquimrocha commented on frontend/src/components/namespace/CreateNamespaceButton.tsx:
```diff
@@ -0,0 +1,167 @@
+              helperText={!namespaceNameValid && namespaceName.length > 0 ? nameHelperMessage : ''}
+              fullWidth
+              value={namespaceName}
+              onChange={event => setNamespaceName(event.target.value)}
```
We could lowercase the name of the namespace here. So users would have it automatically done.

joaquimrocha reviewed (approved):
Almost there. Left a few comments after trying it out.
Seems also like some storybook snapshots that are unrelated did get updated. So I am not sure if it's a legitimate side-effect or if we're now breaking the snapshot generation with this addon.

joaquimrocha reviewed (changes requested):
Approved by mistake. I meant it needs changes.

vyncent-t commented:
Still can't seem to source that build error?

vyncent-t commented:
rebased to current main and updated the `"@types/lodash": "^4.17.4",` to avoid compatibility error, its now `"@types/lodash": "4.17.4",`

joaquimrocha commented on frontend/src/components/namespace/CreateNamespaceButton.tsx:
```diff
@@ -0,0 +1,167 @@
+
+    async function namespaceRequest() {
+      try {
+        const response = await Namespace.apiEndpoint.post(newNamespaceData, {}, clusterData || '');
```
If this succeeds we should set the namespace as '', so if we press the button again, we don't see the name of the namespace we just created.

joaquimrocha commented on frontend/src/components/namespace/CreateNamespaceButton.tsx:
```diff
@@ -0,0 +1,167 @@
+export default function CreateNamespaceButton() {
+  const { t } = useTranslation(['glossary', 'translation']);
+  const [namespaceName, setNamespaceName] = useState('');
+  const [namespaceNameValid, setNamespaceNameValid] = useState(false);
```
It's a good practice to prefix boolean variables about state with `is`. `isNamespaceNameValid`.

joaquimrocha commented on frontend/src/components/namespace/CreateNamespaceButton.tsx:
```diff
@@ -0,0 +1,167 @@
+  }, [namespaceName]);
+
+  return (
+    <>
```
The name needs to be wrapped in a AuthVisible component, because we don't want to show it unless the user can create the namespace.

joaquimrocha reviewed (changes requested):
Almost there. Found a few things in the code.

joaquimrocha commented on frontend/src/components/namespace/CreateNamespaceButton.tsx:
```diff
@@ -0,0 +1,168 @@
+  }
+
+  useEffect(() => {
+    setIsValidNamespaceName(Namespace.isValidNamespaceFormat(namespaceName));
+
+    if (!Namespace.isValidNamespaceFormat(namespaceName)) {
```
These 2 lines call the same function, please set the result of isValidNamespaceFormat to a const and reuse it.

vyncent-t replied:
done!

joaquimrocha reviewed (changes requested):
Just one last change needed.
